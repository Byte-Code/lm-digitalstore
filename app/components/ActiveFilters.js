import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { List } from 'immutable';
import glamorous from 'glamorous';

import Filter from './Filter';

const Wrapper = glamorous.div({
  backgroundColor: 'transparent',
  display: 'flex',
  justifyContent: 'flex-start',
  alignItems: 'center',
  padding: '0 20px',
  width: '100%',
  '&>p': {
    fontFamily: 'LeroyMerlinSans Italic',
    fontSize: '16px',
    color: 'white',
    marginLeft: '28%',
    marginTop: '2%'
  },
  '&>div': {
    marginRight: '10px'
  }
});

export const OtherFilters = glamorous(Filter)({
  color: '#67cb33',
  backgroundColor: '#efefef'
});

export const ResetButton = glamorous.div({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  marginLeft: 'auto',
  cursor: 'pointer',
  '&>p': {
    fontSize: '16px',
    textDecoration: 'underline',
    color: '#fff'
  }
});

const ActiveFilters = props => {
  const {
    filterMap,
    filterGroups,
    resetFilters,
    toggleFilter,
    toggleAvailability,
    handleOpen
  } = props;

  const activeFilters = filterMap.get('filters');
  const activeAvailability = filterMap.get('availability');

  if (activeFilters.isEmpty() && !activeAvailability) {
    return (
      <Wrapper onClick={handleOpen}>
        <p>Tocca per avviare una ricerca avanzata dei prodotti</p>
      </Wrapper>
    );
  }

  const toTake = activeAvailability ? 2 : 3;
  const activeFilterList = filterGroups
    .reduce(
      (acc, val) => acc.push(val.get('filters').filter(f => activeFilters.includes(f.get('code')))),
      List()
    )
    .filterNot(g => g.isEmpty())
    .flatten(true)
    .take(toTake);

  return (
    <Wrapper>
      {activeFilterList.map(f => (
        <Filter key={f.get('code')} isActive onClick={() => toggleFilter(f.get('code'))}>
          <p>{f.get('name')}</p>
        </Filter>
      ))}
      {activeAvailability &&
        <Filter isActive onClick={toggleAvailability}>
          <p>Disponibile</p>
        </Filter>}
      {activeFilters.size > toTake &&
        <OtherFilters width="178px" onClick={handleOpen}>
          <p>Visualizza altri filtri</p>
        </OtherFilters>}
      <ResetButton onClick={resetFilters}>
        <p>Reset filtri</p>
      </ResetButton>
    </Wrapper>
  );
};

ActiveFilters.propTypes = {
  filterGroups: ImmutablePropTypes.list.isRequired,
  resetFilters: PropTypes.func.isRequired,
  handleOpen: PropTypes.func.isRequired,
  toggleFilter: PropTypes.func.isRequired,
  filterMap: ImmutablePropTypes.map.isRequired,
  toggleAvailability: PropTypes.func.isRequired
};

export default ActiveFilters;
