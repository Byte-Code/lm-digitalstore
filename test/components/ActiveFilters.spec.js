import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';

import ActiveFilters, { OtherFilters } from '../../app/components/ActiveFilters';
import Filter from '../../app/components/Filter';

describe('ActiveFilters', () => {
  const toggleFilter = () => {};
  const toggleAvailability = () => {};
  const handleOpen = () => {};
  const filterGroups = fromJS([
    {
      filters: [
        { code: 'filter0', products: [0, 1, 2] },
        { code: 'filter1', products: [3, 4, 5] },
        { code: 'filter2', products: [0, 1, 2] }
      ]
    },
    {
      filters: [
        { code: 'filter3', products: [0, 1, 2] },
        { code: 'filter4', products: [3, 4, 9] },
        { code: 'filter5', products: [6, 7, 8] }
      ]
    }
  ]);

  it('should render a paragrapgh when no filters are active', () => {
    const filterMap = fromJS({
      filters: [],
      availability: null
    });
    const result = shallow(
      <ActiveFilters
        filterMap={filterMap}
        filterGroups={filterGroups}
        toggleFilter={toggleFilter}
        toggleAvailability={toggleAvailability}
        handleOpen={handleOpen}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should render any filters', () => {
    const filterMap = fromJS({
      filters: ['filter0'],
      availability: true
    });
    const result = shallow(
      <ActiveFilters
        filterMap={filterMap}
        filterGroups={filterGroups}
        toggleFilter={toggleFilter}
        toggleAvailability={toggleAvailability}
        handleOpen={handleOpen}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should call toggleFilter when a filter is clicked', () => {
    const filterMap = fromJS({
      filters: ['filter0'],
      availability: false
    });
    const toggleFilterMock = jest.fn();
    const result = shallow(
      <ActiveFilters
        filterMap={filterMap}
        filterGroups={filterGroups}
        toggleFilter={toggleFilterMock}
        toggleAvailability={toggleAvailability}
        handleOpen={handleOpen}
      />
    );
    expect(result).toMatchSnapshot();
    result.find(Filter).simulate('click');
    expect(toggleFilterMock).toHaveBeenCalled();
  });

  it('should call toggleAvailability when a availability filter is clicked', () => {
    const filterMap = fromJS({
      filters: [],
      availability: true
    });
    const toggleAvailabilityMock = jest.fn();
    const result = shallow(
      <ActiveFilters
        filterMap={filterMap}
        filterGroups={filterGroups}
        toggleFilter={toggleFilter}
        toggleAvailability={toggleAvailabilityMock}
        handleOpen={handleOpen}
      />
    );
    expect(result).toMatchSnapshot();
    result.find(Filter).simulate('click');
    expect(toggleAvailabilityMock).toHaveBeenCalled();
  });

  it('should call handleOpen when showMore filter is clicked', () => {
    const filterMap = fromJS({
      filters: ['filter0', 'filter1', 'filter2', 'filter3'],
      availability: false
    });
    const handleOpenMock = jest.fn();
    const result = shallow(
      <ActiveFilters
        filterMap={filterMap}
        filterGroups={filterGroups}
        toggleFilter={toggleFilter}
        toggleAvailability={toggleAvailability}
        handleOpen={handleOpenMock}
      />
    );
    expect(result).toMatchSnapshot();
    result.find(OtherFilters).simulate('click');
    expect(handleOpenMock).toHaveBeenCalled();
  });
});
