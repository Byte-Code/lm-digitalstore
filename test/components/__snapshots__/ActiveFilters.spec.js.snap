// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveFilters should call handleOpen when showMore filter is clicked 1`] = `
<glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p />
  </glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p />
  </glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p />
  </glamorous(div)>
  <glamorous(glamorous(div))
    onClick={[Function]}
    width="178px"
  >
    <p>
      Visualizza altri filtri
    </p>
  </glamorous(glamorous(div))>
</glamorous(div)>
`;

exports[`ActiveFilters should call toggleAvailability when a availability filter is clicked 1`] = `
<glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p>
      Disponibile
    </p>
  </glamorous(div)>
</glamorous(div)>
`;

exports[`ActiveFilters should call toggleFilter when a filter is clicked 1`] = `
<glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p />
  </glamorous(div)>
</glamorous(div)>
`;

exports[`ActiveFilters should render a paragrapgh when no filters are active 1`] = `
<glamorous(div)
  onClick={[Function]}
>
  <p>
    Tocca per avviare una ricerca avanzata dei prodotti
  </p>
</glamorous(div)>
`;

exports[`ActiveFilters should render any filters 1`] = `
<glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p />
  </glamorous(div)>
  <glamorous(div)
    isActive={true}
    onClick={[Function]}
  >
    <p>
      Disponibile
    </p>
  </glamorous(div)>
</glamorous(div)>
`;
